iscservice/demo/main.c

```diff
@@ -0,0 +1,578 @@
+		}
+
+		if (plop->pl_type == M0_LAT_DONE)
+			break;
+
+		if (plop->pl_type == M0_LAT_OUT_READ) {
+			/* XXX just to be sure, for now only */
+			M0_ASSERT(prev_plop != NULL &&
+				  prev_plop->pl_type == M0_LAT_READ);
+			continue;
+		}
+
+		M0_ASSERT(plop->pl_type == M0_LAT_READ); /* XXX for now */
+		m0_layout_plop_start(plop);
+
+		iopl = container_of(plop, struct m0_layout_io_plop, iop_base);
+
+		DBG("req=%d goff=%lu segs=%d\n", reqs_nr, iopl->iop_goff,
+		                                 iopl->iop_ext.iv_vec.v_nr);
+		/* Prepare arguments for computation. */
+		rc = input_prepare(&buf, &comp_fid, iopl, &reply_len, op_type);
+		if (rc != 0) {
+			m0_layout_plop_done(plop);
+			fprintf(stderr, "input preparation failed: %d\n", rc);
+			break;
+		}
+		rc = m0util_isc_req_prepare(req, &buf, &comp_fid, iopl,
+					    reply_len);
+		if (rc != 0) {
+			m0_buf_free(&buf);
+			m0_layout_plop_done(plop);
+			m0_free(req);
+			fprintf(stderr, "request preparation failed: %d\n", rc);
+			break;
+		}
+
+		rc = m0util_isc_req_send(req);
+		conn_addr = m0_rpc_conn_addr(iopl->iop_session->s_conn);
+		if (rc != 0) {
+			fprintf(stderr, "error from %s received: rc=%d\n",
+				conn_addr, rc);
+			break;
+		}
+		reqs_nr++;
+	}
+
+	/* wait for all the replies */
+	while (reqs_nr-- > 0)
+		m0_semaphore_down(&isc_sem);
+
+	/* process the replies */
+	m0util_isc_reqs_teardown(req) {
+		iopl = M0_AMB(iopl, req->cir_plop, iop_base);
+		DBG2("req=%d goff=%lu\n", ++reqs_nr, iopl->iop_goff);
+		if (rc == 0 && req->cir_rc == 0) {
+			if (op_type == ICT_PING)
+				out_args = (void*)conn_addr;
+			out_args = output_process(&req->cir_result,
+					  last && m0_list_is_empty(&isc_reqs),
+						  out_args, op_type);
+		}
+		m0_layout_plop_done(req->cir_plop);
+		m0util_isc_req_fini(req);
+		m0_free(req);
+	}
+
+	return rc;
+}
+
+int main(int argc, char **argv)
+{
+	int                    rc;
+	int                    opt;     /* options */
+	struct m0_op          *op = NULL;
+	struct m0_layout_plan *plan;
+	struct m0_uint128      obj_id;
+	struct m0_indexvec     ext;
+	struct m0_bufvec       data;
+	struct m0_bufvec       attr;
+	int                    op_type;
+	int                    unit_sz;
+	int                    units_nr;
+	m0_bcount_t            len;
+	m0_bcount_t            bs;
+	m0_bindex_t            off = 0;
+	struct m0_obj          obj = {};
+
+	prog = basename(strdup(argv[0]));
+
+	while ((opt = getopt(argc, argv, ":v")) != -1) {
+		switch (opt) {
+		case 'v':
+			trace_level++;
+			break;
+		default:
+			fprintf(stderr, "unknown option: %c\n", optopt);
+			usage();
+			break;
+		}
+	}
+
+	if (argc - optind < 3)
+		usage();
+
+	m0util_setrc(prog);
+
+	op_type = op_type_parse(argv[optind]);
+	if (op_type == -EINVAL)
+		usage();
+	if (read_id(argv[optind + 1], &obj_id) < 1)
+		usage();
+	len = atoll(argv[optind + 2]);
+	if (len < 4) {
+		fprintf(stderr, "object length should be at least 4K\n");
+		usage();
+	}
+	len *= 1024;
+
+	m0trace_on = true;
```

Comment:
Don't introduce so many irrelevant things: like control of m0trace on/off.
This is out of scope of this demo.

Reply:
People often ask about it - how to switch off the generation of m0trace.PID files.